ambry-store/src/main/java/com/github/ambry/store/ScanResults.java

```diff
@@ -26,46 +26,124 @@
+  // be added to its value. But what are deltas? This is an example to show the answer.
+  //
+  // Now let's go through some IndexValues as
+  // 1. PUT[ID1]: created at T0-1D, but will expire at T0+2.5H
+  // 2. PUT[ID2]: created at T0-1D, permanent blob
+  // The validSize at T0 should be PUT[ID1] + PUT[ID2] because PUT[ID1] haven't expired yet. But we know at T0+2.5H, the
+  // validSize should be PUT[ID2] because PUT[ID1] is expired. Since we are not going to handle PUT[ID1] again, we need
+  // to put a delta in T0+3H bucket. And it looks like this:
+  //    T0                           T0+1H                   T0+2H                 T0+3H
+  // PUT[ID1]+PUT[ID2]   [              ]    [                  ]    [ -PUT[ID1]     ]   .....
+  // Please notice that the delta is negative. So when we pick any point of time and add the base validSize and all the
+  // delta values in the buckets before this point of time, we will get the correct answer. For instance, validSize of
+  // T0 is PUT[ID1]+PUT[ID2], validSize of T0+1H is PUT[ID1]+PUT[ID2]+delta of T0+1H. And validSize of T0+2H is PUT[ID1]
+  // +PUT[ID2]-PUT[ID1] because the delta of T0+3H bucket is -PUT[ID1].
+  //
+  // As new IndexValue comes in, we have to deal with them and fill the delta for them as well.
+  // 1. DELETE[ID1]: deleted at T0+0.5H
+  // 2. PUT[ID3]: created at T0+1.2H, but will expire at T0+2.8H
+  // 3. DELETE[ID2]: deleted at T0+1.5H
+  // For DELETE[ID1], it adds -PUT[ID1] delta to T0+1H bucket since it makes PUT[ID1] invalid. However, PUT[ID1] will
+  // expire at T0+2.5H and delta has already been added to T0+3H bucket. In order to avoid double adding, we have to add
+  // PUT[ID1] back to T0+3H bucket. PUT[ID3] adds delta in T0+2H bucket but it will expire at T0+2.8H, so it has be to
+  // subtracted in T0+3H bucket. DELETE[ID2] makes PUT[ID2] invalid, so a delta will be added to T0+2H bucket.
+  //    T0                           T0+1H                   T0+2H                             T0+3H
+  // PUT[ID1]+PUT[ID2]   [-PUT[ID1]      ]    [PUT[ID3]-PUT[ID2] ]    [-PUT[ID1]+PUT[ID1]-PUT[ID3] ]   .....
+  //
+  // This is how delta bucket works. We can pick any point of time and add base validSize and all the delta value in those
+  // buckets prior to this point of time and get the answer we need.
 
+  // Container buckets are very close to the example above, except that it doesn't aggregate all containers' valid data,
+  // rather it keeps them in a map. The key of this map is the AccountId, and the value of this map is yet another map,
+  // whose key is the ContainerId and the value is the valid size of this container.
+  // So {@code containerBaseBucket} is base value for all containers and the {@code containerBuckets} is the delta values
+  // on each time bucket.
+  private final Map<String, Map<String, Long>> containerBaseBucket = new HashMap<>();
+  private final NavigableMap<Long, Map<String, Map<String, Long>>> containerBuckets = new TreeMap<>();
```

Comment:
Let's call this `containerDeltaBuckets`.  Btw, what does `bucketCount` comprise? Is bucketCount = 1 baseBucket + number of delta buckets?

Reply:
sure.

So when bucketCount is 1, we only have baseBucket.